cs/src/core/Index/FASTER/FASTERThread.cs

```diff
@@ -52,129 +54,155 @@ internal long InternalContinue(Guid guid)
+        internal void InternalRelease(FasterExecutionContext ctx)
         {
-            Debug.Assert(threadCtx.Value.retryRequests.Count == 0 &&
-                    threadCtx.Value.ioPendingRequests.Count == 0);
-            if (prevThreadCtx.Value != default(FasterExecutionContext))
+            Debug.Assert(ctx.retryRequests.Count == 0 && ctx.ioPendingRequests.Count == 0);
+            if (ctx.prevCtx != null)
             {
-                Debug.Assert(prevThreadCtx.Value.retryRequests.Count == 0 &&
-                    prevThreadCtx.Value.ioPendingRequests.Count == 0);
+                Debug.Assert(ctx.prevCtx.retryRequests.Count == 0 && ctx.prevCtx.ioPendingRequests.Count == 0);
             }
-            Debug.Assert(threadCtx.Value.phase == Phase.REST);
-            threadCtx.DisposeThread();
-            prevThreadCtx.DisposeThread();
+            Debug.Assert(ctx.phase == Phase.REST);
+
             epoch.Release();
-            overflowBucketsAllocator.Release();
         }
 
-        internal void InitLocalContext(Guid token)
+        internal void InitContext(FasterExecutionContext ctx, Guid token)
         {
-            var ctx = 
-                new FasterExecutionContext
+            ctx.phase = Phase.REST;
+            ctx.version = _systemState.version;
+            ctx.markers = new bool[8];
+            ctx.serialNum = 0;
+            ctx.guid = token;
+
+            if (RelaxedCPR)
+            {
+                if (ctx.retryRequests == null)
                 {
-                    phase = Phase.REST,
-                    version = _systemState.version,
-                    markers = new bool[8],
-                    serialNum = 0,
-                    totalPending = 0,
-                    guid = token,
-                    retryRequests = new Queue<PendingContext>(),
-                    readyResponses = new BlockingCollection<AsyncIOContext<Key, Value>>(),
-                    ioPendingRequests = new Dictionary<long, PendingContext>()
-                };
-
-            for(int i = 0; i < 8; i++)
+                    ctx.retryRequests = new Queue<PendingContext>();
+                    ctx.readyResponses = new AsyncQueue<AsyncIOContext<Key, Value>>();
+                    ctx.ioPendingRequests = new Dictionary<long, PendingContext>();
+                }
+            }
+            else
             {
-                ctx.markers[i] = false;
+                ctx.totalPending = 0;
+                ctx.retryRequests = new Queue<PendingContext>();
+                ctx.readyResponses = new AsyncQueue<AsyncIOContext<Key, Value>>();
+                ctx.ioPendingRequests = new Dictionary<long, PendingContext>();
             }
+        }
 
-            threadCtx.Value = ctx;
+        internal void CopyContext(FasterExecutionContext src, FasterExecutionContext dst)
+        {
+            dst.phase = src.phase;
+            dst.version = src.version;
+            dst.markers = src.markers;
```

Comment:
Not sure if this array assignment is right. Don't we want a fresh set of markers for the new context? You seem to be invoking `CopyContext` within `AtomicSwitch` if the dst.version is smaller than current FASTER version. Is it okay to do a shallow copy of array in that case?